src/librustc/session/mod.rs

```diff
@@ -71,7 +73,58 @@ impl Session {
         self.diagnostic().handler().fatal(msg)
     }
     pub fn span_err(&self, sp: Span, msg: &str) {
-        self.diagnostic().span_err(sp, msg)
+        // Conditions for enabling multi-line errors:
+        if !msg.contains("mismatched types") &&
+           !msg.contains("type mismatch resolving") &&
+           !msg.contains("if and else have incompatible types") &&
+           !msg.contains("if may be missing an else clause") &&
+           !msg.contains("match arms have incompatible types") &&
+           !msg.contains("structure constructor specifies a structure of type") {
+            return self.diagnostic().span_err(sp, msg);
+        }
+
+        let first  = Regex::new(r"[( ]expected").unwrap();
+        let second = Regex::new(r" found").unwrap();
+        let third  = Regex::new(
+            r"\((values differ|lifetime|cyclic type of infinite size)").unwrap();
+
+        let mut new_msg = String::new();
+        let mut beg = 0u;
+
+        // Insert `\n` before expected and found.
+        for (pos1, pos2) in  first.find_iter(msg).zip(
+                            second.find_iter(msg)) {
+            // A `(` may be preceded by a space and it should be trimmed
+            new_msg = new_msg +
+                      msg.slice(beg, pos1.0).trim_right() + // prefix
```

Comment:
I had that but it stopped working yesterday. I swapped to `.slice()` to get it working again.